spring-aop-codeguru/src/main/java/org/springframework/aop/interceptor/AsyncExecutionAspectSupport.java

```diff
@@ -0,0 +1,323 @@
+		this.exceptionHandler = SingletonSupplier.of(exceptionHandler);
+	}
+
+
+	/**
+	 * Configure this aspect with the given executor and exception handler suppliers,
+	 * applying the corresponding default if a supplier is not resolvable.
+	 * @since 5.1
+	 */
+	public void configure(@Nullable Supplier<Executor> defaultExecutor,
+			@Nullable Supplier<AsyncUncaughtExceptionHandler> exceptionHandler) {
+
+		this.defaultExecutor = new SingletonSupplier<>(defaultExecutor, () -> getDefaultExecutor(this.beanFactory));
+		this.exceptionHandler = new SingletonSupplier<>(exceptionHandler, SimpleAsyncUncaughtExceptionHandler::new);
+	}
+
+	/**
+	 * Supply the executor to be used when executing async methods.
+	 * @param defaultExecutor the {@code Executor} (typically a Spring {@code AsyncTaskExecutor}
+	 * or {@link java.util.concurrent.ExecutorService}) to delegate to, unless a more specific
+	 * executor has been requested via a qualifier on the async method, in which case the
+	 * executor will be looked up at invocation time against the enclosing bean factory
+	 * @see #getExecutorQualifier(Method)
+	 * @see #setBeanFactory(BeanFactory)
+	 * @see #getDefaultExecutor(BeanFactory)
+	 */
+	public void setExecutor(Executor defaultExecutor) {
+		this.defaultExecutor = SingletonSupplier.of(defaultExecutor);
+	}
+
+	/**
+	 * Supply the {@link AsyncUncaughtExceptionHandler} to use to handle exceptions
+	 * thrown by invoking asynchronous methods with a {@code void} return type.
+	 */
+	public void setExceptionHandler(AsyncUncaughtExceptionHandler exceptionHandler) {
+		this.exceptionHandler = SingletonSupplier.of(exceptionHandler);
+	}
+
+	/**
+	 * Set the {@link BeanFactory} to be used when looking up executors by qualifier
+	 * or when relying on the default executor lookup algorithm.
+	 * @see #findQualifiedExecutor(BeanFactory, String)
+	 * @see #getDefaultExecutor(BeanFactory)
+	 */
+	@Override
+	public void setBeanFactory(BeanFactory beanFactory) {
+		this.beanFactory = beanFactory;
+	}
+
+
+	/**
+	 * Determine the specific executor to use when executing the given method.
+	 * Should preferably return an {@link AsyncListenableTaskExecutor} implementation.
+	 * @return the executor to use (or {@code null}, but just if no default executor is available)
+	 */
+	@Nullable
+	protected AsyncTaskExecutor determineAsyncExecutor(Method method) {
+		AsyncTaskExecutor executor = this.executors.get(method);
+		if (executor == null) {
+			Executor targetExecutor;
+			String qualifier = getExecutorQualifier(method);
+			if (StringUtils.hasLength(qualifier)) {
+				targetExecutor = findQualifiedExecutor(this.beanFactory, qualifier);
+			}
+			else {
+				targetExecutor = this.defaultExecutor.get();
+			}
+			if (targetExecutor == null) {
+				return null;
+			}
+			executor = (targetExecutor instanceof AsyncListenableTaskExecutor ?
+					(AsyncListenableTaskExecutor) targetExecutor : new TaskExecutorAdapter(targetExecutor));
+			this.executors.put(method, executor);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.



**Problem**  
 You are using a `ConcurrentHashMap`, but your usage of `get()`, `null` check, and `put()` may not be thread-safe at lines: **164, 165, and 179**. Two threads can perform this **same check at the same time** and one thread can **overwrite** the value written by the other thread.

**Fix**  
 Consider replacing `put()` with `putIfAbsent()` to help prevent accidental overwriting. `putIfAbsent()` puts the value only if the `ConcurrentHashMap` does not contain the key and therefore avoids overwriting the value written there by the other thread&#39;s `putIfAbsent()`.

**More info**  
 The thread whose value was overwritten still has a reference to that value and returns that reference to its caller at line: **181**. `putIfAbsent()` returns `null` if the value did not exist and returns the value in the map if one already exists.